src/eng_back.c

```diff
@@ -890,6 +890,64 @@ EVP_PKEY *ctx_load_privkey(ENGINE_CTX *ctx, const char *s_key_id,
 	return PKCS11_get_private_key(key);
 }
 
+static int ctx_keygen(ENGINE_CTX *ctx, void *p)
+{
+	int rv = 1;
+	int i;
+	PKCS11_KGEN_ATTRS *kg_attrs = p;
+	PKCS11_SLOT* slot = NULL;
+
+	pthread_mutex_lock(&ctx->lock);
+	/* Delayed libp11 initialization */
+	if (ctx_init_libp11_unlocked(ctx)) {
+		ENGerr(ENG_F_CTX_LOAD_OBJECT, ENG_R_INVALID_PARAMETER);
+		goto done;
+	}
+
+	// Take the first token that has a matching label
+	// TODO: make this more intelligent
+	for (i = 0; i < ctx->slot_count; ++i) {
+		slot = ctx->slot_list + i;
+		if (slot && slot->token && slot->token->initialized &&
+				slot->token->label &&
+					!strncmp(slot->token->label, kg_attrs->token_label, 32)) {
+			break;
+		}
+	}
+
+	if (i == ctx->slot_count) {
+		ctx_log(ctx, 0, "Initialized token with matching label not found...\n");
+		goto done;
+	}
+
+	if (!ctx->force_login) {
```

Comment:
Can't we merge this with the code below? We need to do a login if force_login is set or the token wants a login?
Now we have code duplication and different kind of error handling/logging if the token requests login or we just force it?

Reply:
I added a comment to make it more clear. This is the same pattern used in ctx_load_object in eng_back.c. I don't exactly know what's the rationale behind this force_login. Maybe performance/security reasons?